src/webapp/components/sidebar/SideBar.tsx

```diff
@@ -1,28 +1,62 @@
-import React from "react";
+import React, { useContext, useEffect, useState } from "react";
 import styled from "styled-components";
-import { Box, Button, Typography } from "@material-ui/core";
+import { Box, Button, CircularProgress, Typography } from "@material-ui/core";
 import { CustomCard } from "../custom-card/CustomCard";
 import { glassColors } from "../../pages/app/themes/dhis2.theme";
-import SidebarNav from "../sidebar-nav/SidebarNav";
+import SidebarNav, { Menu } from "../sidebar-nav/SidebarNav";
 import i18n from "../../../locales";
 import { NavLink } from "react-router-dom";
-import sideBarData from "./sidebar-list.json";
-import { mapModuleToMenu } from "../../hooks/useSidebarMenus";
+
+import { useAppContext } from "../../contexts/app-context";
+import { useGlassModules } from "../../hooks/useGlassModules";
+import { SideBarContext } from "../../contexts/sidebar-context";
+import { mapModuleToMenu } from "./mapModuleToMenu";
 
 export const SideBar: React.FC = () => {
-    const menusResult = { kind: "loaded" as const, data: sideBarData.sideBarData.map(mapModuleToMenu) };
+    const { compositionRoot } = useAppContext();
+    const { setMenuData } = useContext(SideBarContext);
+
+    const [isLoaded, setIsLoaded] = useState(false);
+
+    const storedMenuData: Menu[] | null = JSON.parse(localStorage.getItem("glassSideBarData") || "false") || null;
+
+    const modulesResult = useGlassModules(compositionRoot);
+
+    useEffect(() => {
+        // Validate localstorage vs datastore
+        if (isLoaded && modulesResult.kind === "loaded") {
+            if (modulesResult.data.length) {
+                const dsData = modulesResult.data.map(mapModuleToMenu);
+                if (JSON.stringify(dsData) === JSON.stringify(storedMenuData)) {
```

Comment:
1. JSON.stringify does not guarantee order of keys, So, if the objects to be compared have properties entered in the same order, comparison will work just fine, but in the latter case, where the order has changed, the equality fails. Would it be a better idea to use lodash for equality comparison as we already use the package? 
2. Since we are not doing anything if they are equal, can we just check if (JSON.stringify(dsData) !== JSON.stringify(storedMenuData)) and add to localstorage?